### ISSUE TITLE
GIF - Significant delay before showing the second frame for the very first loop.

### ISSUE THREAD

**Issue by vishalpawale** [bumptech/glide#1060]
**Glide Version:** 3.6.1
**Issue details:**
While loading GIF's there is significant delay before showing **only** the second frame, and this happens only for the first loop, all the later loops goes with expected delay.

So after doing some research, I thought it may be because Glide is resizing the frames, so I override resize operation using `override()`.

**Below is the Glide load line:**
`Glide.with(this).load(trainingItem.getUrl()).asGif().override(Target.SIZE_ORIGINAL, Target.SIZE_ORIGINAL).diskCacheStrategy(DiskCacheStrategy.SOURCE).into(draweeView);`

But still there is significant delay before showing the next frame for the very first loop.
Any suggestions/pointers?

**Comment by TWiStErRob** [comment]
GIF frames are decoded in the same pool as normal images. Are you displaying multiple GIFs or having a lot of other images loaded at the same time?

**Comment by vishalpawale** [comment]
No, I'm showing only one GIF in the whole activity. And even the view hierarchy of that activity is very simple, not so deep. GIF size is around 96KB. If I use high res GIF sizing around 500KB, delay goes upto 20s.

**Comment by TWiStErRob** [comment]
Hmm, that's a lot, those kinds of delays occurred when people are using RESULT cache, which you're not.

I just noticed you call it `draweeView`, are you loading into a standard `ImageView`? If not, please try that.

**Comment by vishalpawale** [comment]
Also, going through code, it looks like we have 30ms as min frame time, and 100ms is default. 

Also it looks to me that we don't have any provision to set maximum frame timing, but in case if we have that provision, I'll set that max frame time to 4-5s and get a workaround for this issue.

**Comment by vishalpawale** [comment]
Yes, I'm loading into standard `ImageView`. 
Please ignore the naming for now, initially I was using Fresco library, but later found Glide really cool. Awesome job man, Thanks for the Glide. :)

**Comment by TWiStErRob** [comment]
Yep, too fast animation is not supported, intentionally, by Glide. I think that min frame time is hardcoded and not configurable. If you want smooth, use a video or an Android animation. Though you're right that if the GIF file contains frame delay value that's 20 seconds Glide will wait.

It's hard to imagine there would be difference in delay on first and other playbacks on purpose. Can you share the image or the whole demo app?

**Comment by vishalpawale** [comment]
Yes, min frame time is not configurable. And I can not go for video or an Android animation, since its a big change on server side as well.

Now please find below the link to the video recording meant for the explanation:
[Dropbox Link](https://www.dropbox.com/s/00o5yzpidtikd8l/Glide%20-%20GIF%20Issue.mp4?dl=0)

**Comment by TWiStErRob** [comment]
Based on this video you don't need to switch to Android anim, because you're not playing at 30fps+ speeds, more like 0.3fps :) So the thumbnail (in the list) and the full-screen view both display 8 seconds on not-first showing. When you opened it showed for 14 seconds. After tapping the thumbnail, the full-screen image loads immediately, which means it was in memory cache. From the code above you are not doing any `.thumbnail()` trick to show a static image while the full GIF animation is loading, so it's likely not that a static image is shown for 6 seconds, and then the real first frame is shown for 8 seconds adding up to 14 seconds.

Can you share that GIF image with us? It's possible there was some timer that is not reset if GifDrawable is reused from memory cache or something like that. Do you know what's the frame delay between the first and second frames?

**Comment by TWiStErRob** [comment]
(Add reference: may be related to #899)

**Comment by vishalpawale** [comment]
Apologies for taking ages to revert back, got busy with something else.
So, after loading that GIF in some browsers, I can observe its taking around 14-16 seconds for the very first loop in IE while chrome and firefox takes 7-8 seconds. Also regarding same I came across [this link](http://webmasters.stackexchange.com/questions/26994/why-is-this-gifs-animation-speed-different-in-firefox-vs-ie)
Attaching the GIF for your reference.
![41a98ef0-2e5f-4ed1-a259-e417a2ec3abf](https://cloud.githubusercontent.com/assets/3822339/14021484/b8814c80-f201-11e5-93b0-b50076381286.gif)

**Comment by TWiStErRob** [comment]
Based on ezgif.com's analyzis this GIF has 600/100=6 seconds delay for each frame encoded in it and has 5 frames. Those GIF researches are not related to this issue, they deal with delays less than 10/100=0.1 seconds=10 milliseconds.

If you have control over those animations, or know someone who has I suggest to add a little overlay saying 1/5, 2/5, ... on each frame, so it's clearer it's an animation, and communicates "there's more to come" for the user.

If you enable `DecodeJob` [logging](https://github.com/bumptech/glide/wiki/Debugging-and-Error-Handling) you can see that the right frames are actually loaded, but it's always the next frame that the log shows. So it looks like GifDrawable displays `firstFrame, 0, 1, 2, 3, 4, 0, 1`, you see the duplicaiton in the beginning. I'm investigating more.

**Comment by TWiStErRob** [comment]
@sjudd I think this code may be wrong:
```java
// GifFrameLoader.loadNextFrame
gifDecoder.advance();
long targetTime = SystemClock.uptimeMillis() + gifDecoder.getNextDelay();
```
I confirmed in debug:

(before/after in table is `gifDecoder.framePointer` and `gifDecoder.getNextDelay()`)

| display | before advance | after advance |
|---|---|---|
| `firstFrame` | -1 -1 | 0 6000 |
| 0 | 0 6000 | 1 6000 |
| 1 | 1 6000 | 2 6000 |

This means that the first frame is displayed twice on the first load and that the frame delays are taken from the wrong frame. As I understand each frame's delay (`GifFrame.delay`) is the amount of time while the frame in question should be on display, that is time to next frame. Because the code calls `advance()` first and then gets the delay frame `n` will be displayed for the time that should be between `n+1` and `n+2`.
Likely no-one has noticed these before because most of the GIF as fast and all frames have the same delay.

I made a version of Vishal's image, that has the following delays (in 1/100th seconds):
![image](https://cloud.githubusercontent.com/assets/2906988/14031408/fbc5a378-f20c-11e5-889c-65e6547f4dc9.png)
So as you can see this image in your browser the "Recalculate fare (grayed popup)" and the "Collect cash (green)"  frames should be only visible for 1 second. However, if you load it in Glide it'll display the two longest (4 second) frames for 1 second.
![ezgif-2670464091](https://cloud.githubusercontent.com/assets/2906988/14031433/1bfc4106-f20d-11e5-94c6-861bbf3f4991.gif)

**Comment by TWiStErRob** [comment]
Fixed in #1091, should available as [3.8.0-SNAPSHOT](https://github.com/bumptech/glide/wiki/Snapshots).

**Comment by vishalpawale** [comment]
Thanks. :)
